### ISSUE TITLE
YOLOv11 Segmentation Masks have artifacts?

### ISSUE THREAD

**Issue by phsilvarepo** [ultralytics/ultralytics#22087]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello there, 

I exported a custom yolov11 model to TFJS format and applied post processing to the output tensors to extract the masks, but currently although the masks are "correct" I keep getting artifacts in the edges of the mask. Is this normal for the YOLO architectures? And what are the steps to remove these artifacts? Finding the biggest blob? Thresholding? Any help would be great.

To extract the masks I performed a sum of the multiplication of the mask coefficients from the Detection output tensor to the proto values in the Segmentation output tensor. Then applied Sigmoid and scaled to match the image original size.

![Image](https://github.com/user-attachments/assets/45199bc6-1881-4f4d-a0d0-89b0833016cb)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @phsilvarepo, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will follow up soon to assist further.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training or export ❓ Question (e.g., TFJS masks post-processing), please share:
- Your exact export command and environment details
- A small code snippet of your post-processing
- One or two sample images plus the resulting visualizations
- Any logs or warnings, and the model details (e.g., YOLO11 variant, image size, training settings)
Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) 📈.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Does the inference with Ultralytics also have the same issue?

**Comment by phsilvarepo** [comment]
Thanks for your reply @Y-T-G, 

No, the inference using Ultralytics does not display this, I think I have to do some other post processing on my package. Below on the blue border is Ultralytics version, on the red the output from my package running on TensorFlow.js a Node.js container.

```
async extractSegmentationMask(maskArr, coeffs, x1, y1, x2, y2, origWidth, origHeight, padTop, padLeft, scale) {
  const proto = tf.tensor(maskArr);                 // [160,160,32]
  const coeff = tf.tensor(coeffs).reshape([1,1,32]);

  let mask = proto.mul(coeff).sum(2);
  mask = tf.sigmoid(mask);

  const x1p = Math.max(0, Math.floor((x1 * scale + padLeft) / 640 * 160));
  const y1p = Math.max(0, Math.floor((y1 * scale + padTop) / 640 * 160));
  const x2p = Math.min(160, Math.ceil((x2 * scale + padLeft) / 640 * 160));
  const y2p = Math.min(160, Math.ceil((y2 * scale + padTop) / 640 * 160));

  const sliceWidth = Math.max(1, x2p - x1p);
  const sliceHeight = Math.max(1, y2p - y1p);

  mask = mask.slice([y1p, x1p], [sliceHeight, sliceWidth]);

  const boxWidth = Math.ceil(x2 - x1);
  const boxHeight = Math.ceil(y2 - y1);
  mask = tf.image.resizeBilinear(mask.expandDims(-1), [boxHeight, boxWidth]).squeeze();

  const maskArray = await mask.array();
  tf.dispose([proto, coeff, mask]);

  return maskArray;
}
```

![Image](https://github.com/user-attachments/assets/6b0dabee-76c5-4ad7-a32f-3b9573a9f088)

![Image](https://github.com/user-attachments/assets/54fc09c8-ee52-45ba-8f4d-10fe56cdfdbc)

**Comment by Y-T-G** [comment]
Ultralytics postprocessing doesn't apply sigmoid

https://github.com/ultralytics/ultralytics/blob/b5fb207d7af1a52103d0d2f146a3b8d2b7e548da/ultralytics/utils/ops.py#L558

**Comment by phsilvarepo** [comment]
Ok but even without sigmoid, although less common it still occurs. If I select a higher threshold it probably erases these artifacts but even with a low threshold the object boundaries compared to Ultralytics are eroded. 

Mine without threshold and no sigmoid:
![Image](https://github.com/user-attachments/assets/f38a7f41-04cb-4712-b5bb-7a98f78f1599)

Mine with 0.01 threshold and no sigmoid:
![Image](https://github.com/user-attachments/assets/c1822184-c04a-431b-824f-ebfd93407295)

Mine with sigmoid:
![Image](https://github.com/user-attachments/assets/5ed339e0-aa48-4933-bfaa-fd627be60fdc)

Ultralytics:
![Image](https://github.com/user-attachments/assets/78863137-32c8-4462-ad56-5bc2af33c637)

**Comment by glenn-jocher** [comment]
Thanks for the clear comparison—this looks like a post‑processing mismatch rather than a model issue. To match Ultralytics YOLO11, mirror `ops.process_mask()` exactly: use boxes in model‑input coordinates, scale them to the proto grid by `(mw/iw, mh/ih)`, crop with the boolean window used in `crop_mask()` (avoid slicing/rounding), optionally upsample with bilinear `align_corners=False`, then binarize at `> 0.0`; only after that undo letterbox with `scale_masks(..., padding=True)` to the original image. References: [process_mask()](https://docs.ultralytics.com/reference/utils/ops/#process_mask), [crop_mask()](https://docs.ultralytics.com/reference/utils/ops/#crop_mask), [scale_masks()](https://docs.ultralytics.com/reference/utils/ops/#scale_masks), and the TF.js export outputs (boxes + protos) in the [exporter forward reference](https://docs.ultralytics.com/reference/engine/exporter/#forward). If it still diverges, please share one sample of raw exported `protos`, `coeffs`, and the exact `boxes` you use (in their coordinate space) so we can reproduce.

**Comment by RSWAIN1486** [comment]
> Thanks for your reply [@Y-T-G](https://github.com/Y-T-G),
> 
> No, the inference using Ultralytics does not display this, I think I have to do some other post processing on my package. Below on the blue border is Ultralytics version, on the red the output from my package running on TensorFlow.js a Node.js container.
> 
> ```
> async extractSegmentationMask(maskArr, coeffs, x1, y1, x2, y2, origWidth, origHeight, padTop, padLeft, scale) {
>   const proto = tf.tensor(maskArr);                 // [160,160,32]
>   const coeff = tf.tensor(coeffs).reshape([1,1,32]);
> 
>   let mask = proto.mul(coeff).sum(2);
>   mask = tf.sigmoid(mask);
> 
>   const x1p = Math.max(0, Math.floor((x1 * scale + padLeft) / 640 * 160));
>   const y1p = Math.max(0, Math.floor((y1 * scale + padTop) / 640 * 160));
>   const x2p = Math.min(160, Math.ceil((x2 * scale + padLeft) / 640 * 160));
>   const y2p = Math.min(160, Math.ceil((y2 * scale + padTop) / 640 * 160));
> 
>   const sliceWidth = Math.max(1, x2p - x1p);
>   const sliceHeight = Math.max(1, y2p - y1p);
> 
>   mask = mask.slice([y1p, x1p], [sliceHeight, sliceWidth]);
> 
>   const boxWidth = Math.ceil(x2 - x1);
>   const boxHeight = Math.ceil(y2 - y1);
>   mask = tf.image.resizeBilinear(mask.expandDims(-1), [boxHeight, boxWidth]).squeeze();
> 
>   const maskArray = await mask.array();
>   tf.dispose([proto, coeff, mask]);
> 
>   return maskArray;
> }
> ```
> 
> ![Image](https://github.com/user-attachments/assets/6b0dabee-76c5-4ad7-a32f-3b9573a9f088)
> 
> ![Image](https://github.com/user-attachments/assets/54fc09c8-ee52-45ba-8f4d-10fe56cdfdbc)

Hello @phsilvarepo 

I have been looking for a tensorflow js implementation of yolov11 in a node js container. Would you be able to share the repo? Thanks in advance. Cheers!!
